valhalla/meili/topk_search.h

```diff
@@ -32,24 +36,30 @@ class EnlargedTransitionCostModel
   EnlargedTransitionCostModel(const EnlargedViterbiSearch& evs)
       :evs_(evs) {}
 
-  float operator() (const StateId& lhs, const StateId& rhs) const;
+  float operator() (const StateId& lhs, const StateId& rhs);
 
  private:
+  float calculate_cost(const StateId& lhs, const StateId& rhs) const;
+
   const EnlargedViterbiSearch& evs_;
+  std::unordered_map<std::pair<StateId, StateId>, float> cached_costs_;
 };
 
 class EnlargedViterbiSearch
 {
  public:
-  EnlargedViterbiSearch(IViterbiSearch& vs, std::function<StateId(const StateId::Time&)> claim_stateid)
+  EnlargedViterbiSearch(IViterbiSearch& vs, std::function<StateId (const StateId::Time&)> claim_stateid,
+      std::unordered_map<StateId, StateId> &initial_origins, std::unordered_set<StateId>& removed_origins)
       : vs_(vs),
         claim_stateid_(claim_stateid),
         original_emission_cost_model_(vs.emission_cost_model()),
         original_transition_cost_model_(vs.transition_cost_model()),
         origin_(),
         clone_(),
         clone_start_time_(kInvalidTime),
-        clone_end_time_(kInvalidTime)
+        clone_end_time_(kInvalidTime),
+        initial_origins_(initial_origins),
+        removed_origins_(removed_origins)
```

Comment:
`removed_stateids` might be better because it allows removing clones too.

Reply:
only original ids are ever removed because those are the ones we use in `RemoveRendundancies` to check for redundant paths. these are candidates that are found within the `StateContainer`.

we also use this trick to remove clones by checking above for the clones original id in the enlarged emission cost model. this way we dont have to store clones at all, we simply look what the original id was and if the redundancy check marked it then the emission cost is `-1` and the clones path will be removed